authstate.go

```diff
@@ -0,0 +1,117 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/matrix-org/util"
+)
+
+// StateProvider is capable of returning the room state at any point in time.
+type StateProvider interface {
+	// StateIDsAtEvent returns a list of state event IDs for the event ID provided, which represent the entire
+	// room state at that event.
+	StateIDsAtEvent(ctx context.Context, roomID, atEventID string) ([]string, error)
```

Comment:
`At` is an unclear term here when referring to state events. It would be better to use `Before` or `After` (where `Before` doesn't include state changes from the event and `After` does include them).

Reply:
I agree, and actually don't know the answer as the spec just says:

> Passes authorization rules based on the state at the event, otherwise it is rejected.

Reply:
I guess `Before`, else if you say changed power levels to give yourself admin then you'd be able to as the state at that point would already be giving yourself admin.